### PR TITLE
Bump actions/artifact-upload to v4

### PR DESCRIPTION
**Pull request by danielhollas**
This should significantly speed up artifact upload, see https://github.com/actions/upload-artifact#v4---whats-new

TODO: Probably we should turn off compression, see https://github.com/actions/upload-artifact#altering-compressions-level-speed-v-size

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -3,53 +3,13 @@ name: Build, test and push Docker Images
 
 on:
     pull_request:
-        paths:
-            - ".github/workflows/docker.yml"
```
Comment (danielhollas) [inline .github/workflows/docker.yml:7]: This seemed to include essentially all files and seemed very fragile. For now I simply excluded Markdown files, although I tempted to even remove that.

File: .github/actions/create-dev-env/action.yml [hunk 2]
```diff
@@ -18,7 +18,7 @@ runs:
           if: ${{ inputs.architecture == 'amd64' }}
           uses: actions/setup-python@v4
           with:
-              python-version: 3.x
+              python-version: 3.11
```
[the remark this answers is not in the document]
Comment (unkcpz) [reply .github/actions/create-dev-env/action.yml:21]: Yep, agree.

**Comment by unkcpz** [comment]
Thanks! I just noticed the update. It makes upload much faster! The test failed because of Marnik's machine was shutdown, now should build and test on arm64.

> Probably we should turn off compression

How much time it will save without compress?

File: .github/workflows/docker.yml (hunk 1 above)
Comment (unkcpz) [reply .github/workflows/docker.yml:7]: Makes a lot sense for this repo, go ahead.

**Comment by danielhollas** [comment]
> How much time it will save without compress?

Seems to save a lot. Even the lowest level of compression (lvl 1) is ~20s slower. Note the download-artifact still takes around 1 minute, and there are other extra steps (saving and loading Docker images) that make all this slower.

My apologies, I have by accident pushed some commits on master, so I needed to merge this without approval. Please leave a review if you have any comments and I'll make a new PR.

**Pull request merged**